language-plutus-core/src/Language/PlutusCore/Normalize.hs

```diff
@@ -93,22 +113,22 @@ substituteNormalizeTypeM
     => NormalizedType tyname ()                          -- ^ @ty@
     -> tyname ()                                         -- ^ @name@
     -> Type tyname ()                                    -- ^ @body@
-    -> NormalizeTypeM tyname (NormalizedType tyname ())  -- ^ @NORM ([ty / name] body)@
+    -> NormalizeTypeM tyname a (NormalizedType tyname ())  -- ^ @NORM ([ty / name] body)@
 substituteNormalizeTypeM ty name = withExtendedTypeEnv name ty . normalizeTypeM
 
 -- See Note [Normalization].
 -- | Normalize a 'Type'.
 normalizeType
-    :: (HasUnique (tyname ()) TypeUnique, MonadQuote m)
-    => Type tyname () -> m (NormalizedType tyname ())
-normalizeType = runNormalizeTypeM . normalizeTypeM
+    :: (HasUnique (tyname ()) TypeUnique, MonadQuote m, AsTypeError e a, MonadError e m)
+    => Maybe Natural -> Type tyname () -> m (NormalizedType tyname ())
+normalizeType n = runNormalizeTypeM n . normalizeTypeM
 
 -- See Note [Normalizing substitution].
 -- | Substitute a type for a variable in a type and normalize.
 substituteNormalizeType
-    :: (HasUnique (tyname ()) TypeUnique, MonadQuote m)
+    :: (HasUnique (tyname ()) TypeUnique, MonadQuote m, AsTypeError e a, MonadError e m)
     => NormalizedType tyname ()      -- ^ @ty@
     -> tyname ()                     -- ^ @name@
     -> Type tyname ()                -- ^ @body@
     -> m (NormalizedType tyname ())  -- ^ @NORM ([ty / name] body)@
-substituteNormalizeType ty name = runNormalizeTypeM . substituteNormalizeTypeM ty name
+substituteNormalizeType ty name = runNormalizeTypeM Nothing . substituteNormalizeTypeM ty name
```

Comment:
I think it would be clearer if this also took a gas supply. We want to be careful at all the use sites about whether we should be tracking the gas or not, and I'm not 100% sure we never want to for a call to this function.

Reply:
In fact I'm pretty sure this should be taking gas. In the situation where we're given a program with normalized types on the terms, this is is the only thing that will be doing normalization, so we had better be restricting it!